packages/project-build/src/shared/graph-utils.ts

```diff
@@ -0,0 +1,93 @@
+import type { Instance } from "@webstudio-is/sdk";
+
+type InstanceId = Instance["id"];
+
+// Depth-First Search (DFS) algorithm to find cycles in a directed graph
+export const findCycles = (
+  graph: Iterable<Pick<Instance, "id" | "children">>
+): InstanceId[][] => {
+  const adjacencyList: Record<InstanceId, InstanceId[]> = {};
+
+  // Build adjacency list
+  for (const node of graph) {
+    adjacencyList[node.id] = node.children
+      .filter((child) => child.type === "id")
+      .map((child) => child.value);
+  }
+
+  const visited = new Set<string>();
+  const path: InstanceId[] = [];
+  const cycles: InstanceId[][] = [];
+
+  const dfs = (nodeId: string): void => {
+    if (path.includes(nodeId)) {
+      const cycleStart = path.indexOf(nodeId);
```

Comment:
could be single path.indexOf instead of both includes and insteadOf

Reply:
Usually in this algorithms here is not inlined code but some `normalizePath` useless for our purposes.